Security considerations: Attester and Keys (#138)
ietf-rats-wg/architecture: draft-ietf-rats-architecture.md

# Security Considerations

## Attester and Attestation Key Protection

> (this is a crude first pass to see what interest there is in this)

Reviewer: Yes I think this content is useful, so comments below try to clean it up a bit.  Also some parts below might be more generalizable beyond just the manufacturer, as noted below.

> Implementors need to pay close attention to the isolation and protection of the Attester and the factory processes for provisioning the Attestation Key Material. When these are compromised, the attestation becomes worthless because the attacker can forge attestations.

Reviewer 2: This reads to me as of the Attester should be isolated (e.g. air-gap). What it should read like is that there are components in an Attester that have to be trustworthy due to isolation or shielding. Would others please reply to this comment here and highlight if they agree or not?

> Implementors need to pay close attention to the isolation and protection of the Attester and the factory processes for provisioning the Attestation Key Material. When these are compromised, the attestation becomes worthless because the attacker can forge attestations.

Reviewer 2: ```suggestion
Implementors need to pay close attention to the isolation and protection of the Attester and the factory processes for provisioning the Attestation Key Material. When these factory processes are compromised, the attestation becomes worthless because the attacker can forge attestations.
```

> Attestation applies to use cases with a range of security requirements, so the protections discussed here range from low to high security where low security  may be only application or process isolation by the devices operating system and high security involves specialized HW to defend against physical attacks on a chip.

Reviewer 3: ```suggestion
Remote attestation applies to use cases with a range of security requirements, so the protections discussed here range from low to high security where low security  may be only application or process isolation by the device's operating system and high security involves specialized hardware to defend against physical attacks on a chip.
```

### On-Device Attester and Key Protection

> It is assumed that the Attester is an isolated computing environment like a process, a dedicated chip a TEE or such that collects the Claims, formats them and signs them with the Attestation Key. The Attester must be protected from modification to be sure it behaves correctly. There must also be confidentiality so that the signing key is not captured and used elsewhere to forge attestations.

Reviewer 2: ```suggestion
It is assumed that the Attester is located in an isolated environment of a device like a process, a dedicated chip a TEE or such that collects the Claims, formats them and signs them with an Attestation Key. The Attester must be protected from unauthorized modification to ensure it behaves correctly. There must also be confidentiality so that the signing key is not captured and used elsewhere to forge evidence.
```

Reviewer 2: (did not address grammar issues as highlighted by Dave)

> In many cases the user or owner of the device must not be able to modify or inspect the Attester. For example the owner or user of a mobile phone of FIDO authenticator is not trusted. The point of the attestation is for the relying party to be able to trust the device even though they don’t trust the user or owner. 

Reviewer 2: If find this sentence a bit confusing to read:

Please highlight whyt a user or owner must not be able to modify or inspect an Attester. Do you mean effectively the Attesting Environment of the Attester? That would make way more sense to me.

Why would you want a user/owner of a phone to provide biometric digests, pins or passwords, if you do not trust her in the first place?

"The point of the attestation": I really do not know what that means here. Please remove 'attestation' and elaborate on the goal of evidence generation earlier in the text please. Alternatively, "remote attestation procedure" might already be good enough?

> In many cases the user or owner of the device must not be able to modify or inspect the Attester. For example the owner or user of a mobile phone of FIDO authenticator is not trusted. The point of the attestation is for the relying party to be able to trust the device even though they don’t trust the user or owner. 

Reviewer 3: ```suggestion
In many cases the user or owner of the device must not be able to modify or exfiltrate keys from the Attesting Environment of the Attester. 
For example the owner or user of a mobile phone or FIDO authenticator is not trusted. 
The point of remote attestation is for the Relying Party to be able to trust the Attester even though they don’t trust the user or owner. 
```

> In other cases the user or owner is simply trying to monitor a remote device and it is OK for them to be able to modify or inspect the Attester. For example, the owner of a router may wish to monitor the state of a router in a wiring closet far away.

Reviewer 2: I don't understand what problem is intended to be addressed here in the first place, unfortunately.

> In other cases the user or owner is simply trying to monitor a remote device and it is OK for them to be able to modify or inspect the Attester. For example, the owner of a router may wish to monitor the state of a router in a wiring closet far away.

Reviewer 3: ```suggestion
```

> Some of the measures for low security include process or application isolation by a high-level operating system, perhaps restricting access root or system privilege. For extremely simple single-use devices that don’t use a protected mode operating system, like a sound bar, the isolation might only be the plastic housing for the device.

Reviewer 3: ```suggestion
Some of the measures for low security include process or application isolation by a high-level operating system, perhaps restricting access to root or system privilege. For extremely simple single-use devices that don’t use a protected mode operating system, like a sound bar, the isolation might only be the plastic housing for the device.
```

At medium level security, a special restricted operating environment like a TEE might be used. In this case only security-oriented SW has access to the attester and key material. 

> For high level security, specialized hardware will likely be used providing protection against chip decapping attacks, power supply and clock glitching, faulting injection and RF and power side channel attacks.

Reviewer 2: Now I would have expected a "low level of security" earlier in the text :-)

### Factory Key Provisioning Processes

This refers to the process that occur in the factory or elsewhere that establish the signing key material on the device and the verification key material off the device. Sometimes this is referred to as “personalization”. 

> One way to provision a key is to generate it off the device and copy the secret key on to the device. In this case confidentiality of the path over which the key is provisioned is necessary. This can be achieved in a number of ways.

Reviewer 2: ```suggestion
One way to provision a key is to first generate it external to the device and then copy the secret key on to the device. In this case confidentiality of the path over which the key is provisioned is necessary. This can be achieved in a number of ways.
```

> It can be achieved entirely with physical factory security involving no encryption at all. For low-security use cases this might be simply locking doors and limiting personal that can enter the factory. For high-security use cases this might involve a special area of the factory accessible only to select, security trained personnel.

Reviewer 2: ```suggestion
Confidentiality can be achieved entirely with physical factory security involving no encryption at all. For low-security use cases this might be simply locking doors and limiting personal that can enter the factory. For high-security use cases this might involve a special area of the factory accessible only to select, security trained personnel.
```

> Cryptography can also be used to help implement the confidentiality, but note that the keys used to provision the attestation key must somehow have been provisioned confidentially before.

Reviewer 2: ```suggestion
Cryptography can also be used to support confidentiality, but secret keys that are used to then provision attestation keys must somehow have been provisioned securely beforehand (a recursive problem).
```

In many cases both some physical security and some cryptography will be necessary and useful to establish confidentiality.

> Another way to provision the key material is to generate it on the device and export the verification key. If public key cryptography is being used, then only integrity is necessary. Confidentiality is not necessary.

Reviewer: Comment: a lot of this section would also apply to the process by which an owner-generated (or other third-party generated) certificate is generated for the device.  This is important for cases where the verifier uses endorsements from a part other than the manufacturer, whether instead of or in addition to the manufacturer's.

> In all cases, the factory process must be sure that it only establishes key material in Attesters that are made by the manufacturer and configured correctly. For many use cases, this will involve physical security at the factory, for example to make sure an attacker doesn’t do a “midnight” production run of some counterfeit or incorrectly configured devices.

Reviewer 2: ```suggestion
In all cases, a factory process must ensure that only key material that is generated by the manufacturer is establishes in Attesters and then configured correctly. For many use cases, this will involve physical security at the factory, for example to ensure an attacker doesn’t do a “midnight” production run of some counterfeit or incorrectly configured devices.
```

Reviewer 2: I think I know what a 'midnight run' is. Does everybody else?

# Security Considerations

## Attester and Attestation Key Protection

> (this is a crude first pass to see what interest there is in this)

Reviewer 3: ```suggestion
```

In many cases the user or owner of the device must not be able to modify or exfiltrate keys from the Attesting Environment of the Attester. 
For example the owner or user of a mobile phone or FIDO authenticator is not trusted. 
The point of remote attestation is for the Relying Party to be able to trust the Attester even though they don’t trust the user or owner. 

Some of the measures for low level security include process or application isolation by a high-level operating system, and perhaps restricting access to root or system privilege. For extremely simple single-use devices that don’t use a protected mode operating system, like a Bluetooth speaker, the isolation might only be the plastic housing for the device.

At medium level security, a special restricted operating environment like a Trusted Execution Environment (TEE) might be used. In this case, only security-oriented software has access to the Attester and key material. 

For high level security, specialized hardware will likely be used providing protection against chip decapping attacks, power supply and clock glitching, faulting injection and RF and power side channel attacks.

> ### Factory Key Provisioning Processes
> 
> Factory key provisioning is the process that occurs in the factory or elsewhere that establishes the signing key material on the device and the verification key material off the device. Sometimes this is referred to as “personalization”. 

Reviewer 3: ```suggestion
### Attestation Key Provisioning Processes

Attestation key provisioning is the process that occurs in the factory or elsewhere that establishes the signing key material on the device and the verification key material off the device. Sometimes this is referred to as “personalization”. 
```

One way to provision a key is to first generate it external to the device and then copy the key onto the device. In this case, confidentiality of the generator, as well as the path over which the key is provisioned, is necessary. This can be achieved in a number of ways.

> Confidentiality can be achieved entirely with physical factory security involving no encryption at all. For low-security use cases, this might be simply locking doors and limiting personnel that can enter the factory. For high-security use cases, this might involve a special area of the factory accessible only to select security-trained personnel.

Reviewer 3: ```suggestion
Confidentiality can be achieved entirely with physical provisioning environment security involving no encryption at all. For low-security use cases, this might be simply locking doors and limiting personnel that can enter the factory. For high-security use cases, this might involve a special area of the factory accessible only to select security-trained personnel.
```

Cryptography can also be used to support confidentiality, but keys that are used to then provision attestation keys must somehow have been provisioned securely beforehand (a recursive problem).

In many cases both some physical security and some cryptography will be necessary and useful to establish confidentiality.

Another way to provision the key material is to generate it on the device and export the verification key. If public key cryptography is being used, then only integrity is necessary. Confidentiality is not necessary.

> In all cases, the Attestation Key provisioning process must be sure that it only establishes key material in Attesters that are made by the manufacturer and configured correctly. 

Reviewer 3: ```suggestion
In all cases, the facility process must ensure that only attestation key material that is generated by a valid Endorser is established in Attesters and then configured correctly.
```

> In all cases, the factory process must be sure that it only establishes key material in Attesters that are made by the manufacturer and configured correctly. For many use cases, this will involve physical security at the factory, for example to make sure an attacker doesn't do a "midnight" production run of some counterfeit or incorrectly configured devices.

Reviewer 2: ```suggestion
In all cases, the factory process must ensure that only key material that is generated by the manufacturer is established in Attesters and then is configured correctly. For many use cases, this will involve physical security at the factory, for example to ensure an attacker can not do a "midnight" production run of some counterfeit or incorrectly configured devices.
```

> In all cases, the factory process must be sure that it only establishes key material in Attesters that are made by the manufacturer and configured correctly. For many use cases, this will involve physical security at the factory, for example to make sure an attacker doesn't do a "midnight" production run of some counterfeit or incorrectly configured devices.

Reviewer 3: ```suggestion
In all cases, the Attestation Key provisioning process must be sure that it only establishes key material in Attesters that are made by the manufacturer and configured correctly. 
For many use cases, this will involve physical security at the facility, to prevent unauthorized devices from being manufactured that may be counterfeit or incorrectly configured.
```
